### DISCUSSION TITLE
Execute DDL statements in parallel

### DISCUSSION THREAD

**Discussion by coenvk** [TPC-Council/HammerDB#322] [Q&A] [answered]
Hi!

This is my first time using HammerDB and I am trying to setup a test for a PostgreSQL database. I have already been able to use HammerDB with a simple setup.

However, I would like to run a custom script in parallel as well. I am using a schema versioning tool called Liquibase. Liquibase allows me to run a command like: `liquibase update` which executes any new schema changes.

What I would like to do:
- Build the schema for HammerDB
- Start the virtual users with a ramp-up time
- After the ramp-up, put a load on the database for say 20 minutes
- After 10 minutes of pressure from the load, execute a schema change
- Observe the behavior for the remaining 10 minutes of load

Is there a way to do this?

Additionally, I would like to output some metrics (such as the latency and throughput) to a file, so I can create a chart out of them. How would I go about that?

Thank you for your help in advance!

**Comment by sm-shaw** [comment, marked as answer]
Hi,
HammerDB is multithreaded, and the virtual users already run in parallel.  Also, it is scripted for exactly this reason, so you change the test scenario however you wish. 

So you will want to be running a timed test and here virtual user 1 is the monitor virtual user. So for example in the GUI look in the script editor that is loaded for PostgreSQL and find the following timing section. In this example, it is modified to print out "running bespoke command after 2 mins" after 2 minutes.  So you can put anything you want in here - to execute an external command use "exec", for SQL commands, see examples later in the driver script. 

```
        puts "Timing test period of $duration in minutes"
            set testtime 0
            set durmin $duration
            set duration [ expr $duration*60000 ]
            while {$testtime != $duration} {
                if { [ tsv::get application abort ] } { break } else { after 6000 }
                set testtime [ expr $testtime+6000 ]
                if { ![ expr {$testtime % 60000} ] } {
                    puts -nonewline  "[ expr $testtime / 60000 ]  ...,"
	      if {[ expr $testtime / 60000 ] eq 2} {
		puts "running bespoke command after 2 mins"
		}
                }
            }
```
This gives output as follows
```
Hammerdb Log @ Thu Mar 10 17:49:15 GMT 2022
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
Vuser 1:Beginning rampup time of 2 minutes
Vuser 2:Processing 10000000 transactions with output suppressed...
Vuser 1:Rampup 1 minutes complete ...
Vuser 1:Rampup 2 minutes complete ...
Vuser 1:Rampup complete, Taking start Transaction Count.
Vuser 1:Timing test period of 4 in minutes
Vuser 1:1 ...,
Vuser 1:2 ...,
Vuser 1:running bespoke command after 2 mins
Vuser 1:3 ...,
Vuser 1:4 ...,
Vuser 1:Test complete, Taking end Transaction Count.

```
If you want to load a custom script in the CLI use the customscript command. note you can write and test it in the GUI and then save it and load it in the CLI for ease of use. 

For latency, see the documentation on time profiling - this is a very comprehensive way to capture all the transaction latencies you need. 
https://www.hammerdb.com/blog/uncategorized/hammerdb-v4-1-extended-time-profiling/
https://www.hammerdb.com/docs/ch04s06.html

For throughput, see logging transaction counter output https://www.hammerdb.com/docs/ch06s08.html

Also you can view PostgreSQL statistics over time with the graphical metrics viewer 
https://www.hammerdb.com/blog/uncategorized/hammerdb-v4-3-new-features-pt1-graphical-metrics-for-postgresql/
This means if you run the metrics viewer over the time of your test you will be able to drag the window over different sections of the output and view the metrics as they were for that particular moment in time. This will allow you to compare before and after.

**Reply by coenvk** [reply]
Hi,
Thank you very much for this quick and comprehensive reaction. This is exactly what I am trying to do and I will definitely try this out.
